### ISSUE TITLE
Perf: extend Coordinate with distanceSq and distance3DSq

### ISSUE THREAD

**Issue by jandam** [locationtech/jts#175]
This is simple optimization to avoid computing Math.sqrt during finding minimal distance.
Suggested method can be used for example in LineSegment, Triangle, InteriorPointPoint, InterirorPointLine, , LineString::equalsExact(Geometry other, double tolerance)

```
 public double distanceSq(Coordinate c) {
    double dx = x - c.x;
    double dy = y - c.y;
    return dx * dx + dy * dy;
  }

  public double distance3DSq(Coordinate c) {
    double dx = x - c.x;
    double dy = y - c.y;
    double dz = z - c.z;
    return dx * dx + dy * dy + dz * dz;
  }
```

**Comment by dr-jts** [comment]
Interesting.  Is the use of `Math.sqrt()` a significant performance hit?

**Comment by jandam** [comment]
I didn't make any benchmarks. It's just very simple micro optimization used in games. Computing min distance is O(n). And we can compute Math.sqrt only once.

**Comment by dr-jts** [comment]
I think I'd prefer to name these `distanceSqr `and `distanceSqr3D`

**Comment by jandam** [comment]
Yes it's better. English is not my native language. So method names are up tu you.

**Comment by jnh5y** [comment]
I'd agree that comparing and minimizing the square of the distances is a typical trick to save some cycles.  I'm +1 on it.

**Comment by FObermaier** [comment]
`DistanceSquared.pointToPoint(Coordinate, Coordinate)` is part of https://github.com/locationtech/jts/pull/427.
But not as a member of `Coordinate`
